### DISCUSSION TITLE
Adding a project for GSoC 2024

### DISCUSSION THREAD

**Discussion by kshitijdshah99** [aiidateam/aiida-core#6300] [Ideas]
I had recently viewed that AiiDA had came up with a very cool project regarding training a Transformer from Scratch can we open up that project for GSoC 2024 AiiDA once again

**Comment by ltalirz** [comment]
@unkcpz

**Comment by unkcpz** [comment]
Hi @kshitijdshah99, thanks for the interested on the project. 

However, after internal discussion, we decide not to list the workchain/querybuilder ML/LLM project. 
We either do not have clear goal on what to achieve nor we don't have sufficient ability to tutoring  the students. 

I put the latest version of the project idea below, if you are interested, the proposal is always welcome. Please notice the deadline of the student application is **2nd April**, and before that we are willing to see the potential of student by some small contributions to the community or some online discussion on the topic. 

## Project 2 - Training an LLM to generate workchain and query from natural language prompts 

**level** *advanced*

**Expected Size** 350h

In AiiDA, WorkChain is a robust workflow capable of recovering from interruptions during execution, making it ideal for handling expensive calculations. However, mastering its usage requires familiarity with certain concepts before one can confidently create a resilient workchain.

Same for the `QueryBuilder` one of the most powerful tool in AiiDA to find target information from your huge amout of data generate from calculations.
However, it can be challenging to learn to write a correct query which even sometime not straightforward for experienced users.

This project aims to fine-tuning a large language model (LLM) to allow users to boilerplate WorkChain or build query they are interested in by expressing their desired workflow or data in a few sentences.
Existing LLM's such as ChatGPT already do a fair job at this, but the produced code is still often incorrect and outdated.

Generating a diverse dataset of query/workchain prompts and corresponding Python code, and using this data to improve a dedicated LLM will hopefully make the produced AiiDA queries/workchains more accurate, creating a powerful tool for users to build workchain and extract the results they are interested in.

### Expected outcomes

At the end of the project, we aim to have a lightweight tool that can: 
- Provide a starting template for a `WorkChain` code based on user input, which can then be refined futher.
- Generate a correct `QueryBuilder` instance from a user prompt.

This will require:

- A database that maps natural language prompts to the corresponding queries/workchains, which can be easily and incrementally expanded. 
- LLM fine-tuning on this database that converts prompts into a `QueryBuilder`/`WorkChain` instance.
- A user-friendly interface for that can be installed locally in the form of a Python package and optionally an online tool integrated with the Materials Cloud.

### Skills

- We expect you to be familiar with object-oriented programming in **Python**.
- You need to have experience in natural language processing and LLM.

### Note

This project poses an exciting challenge for both students and mentors. 
While the AiiDA team may not have extensive experience with LLM, we eagerly anticipate students bringing their knowledge to the table. 
We are ready to provide expertise for the actual queries and more, making this collaboration a dynamic and enriching opportunity.

**Reply by kshitijdshah99** [reply]
Hey, @unkcpz that's great I am actually interested in this topic too bcz I have a hands on experience of making  and training a transformer from scratch using numpy and cupy libraries for making a chatbot, English to German translation and also NYT dataset for generating news.
Here's my [repo link](https://github.com/kshitijdshah99/Transformer-From-Scratch).

**Reply by kshitijdshah99** [reply]
Is this project officially listed for GSoC 2024, bcz I could not find it in the [README.md where project 1 was listed](https://github.com/aiidateam/aiida-core/wiki/GSoC-2024-Projects)? 
@unkcpz are you going to mentor this project, if yes then that's great...
I will start exploring regarding the concepts and workflow of this project after which I can start contributing to it. Can we have a meet next week once I am done with exploring this...need to clarify my doubts and finalize the workflow.

**Reply by unkcpz** [reply]
Thanks for interesting on it. For sure we can meet and have a chat. Can we meet at Friday at CET 10 am (Let me know your time zone and we can find a time proper for both of us)?

**Reply by kshitijdshah99** [reply]
Great, @unkcpz can we have it at Thursday CET 5pm i.e. (9:30 pm IST) or next Monday (any time you say) because my exams just got over this week and am I bit occupied in some work this weekend.

**Reply by unkcpz** [reply]
Let's touch base at Thursday CET 5pm. Can you give me your email so I can send you the meeting link?

**Reply by kshitijdshah99** [reply]
[kshitij.d.shah.999@gmail.com](kshitij.d.shah.999@gmail.com)
